matchbox_signaling/src/topologies/client_server.rs

```diff
@@ -0,0 +1,282 @@
+    async fn state_machine(upgrade: WsStateMeta<ClientServerCallbacks, ClientServerState>) {
+        let WsStateMeta {
+            peer_id,
+            sender,
+            mut receiver,
+            mut state,
+            callbacks,
+        } = upgrade;
+
+        // The first person to connect becomes host.
+        if state.get_host().is_none() {
+            // Set host
+            state.set_host(peer_id, sender.clone());
+            // Lifecycle event: On Host Connected
+            callbacks.on_host_connected.emit(peer_id);
+        } else {
+            // Alert server of new user
+            let event = Message::Text(JsonPeerEvent::NewPeer(peer_id).to_string());
+            // Tell host about this new client
+            match state.try_send_to_host(event) {
+                Ok(_) => {
+                    // Add peer to state
+                    state.add_client(peer_id, sender.clone());
+                    // Lifecycle event: On Client Connected
+                    callbacks.on_client_connected.emit(peer_id);
+                }
+                Err(e) => {
+                    error!("error sending peer {peer_id:?} to host: {e:?}");
+                    return;
+                }
+            }
+        }
+
+        // Check whether this connection is host
+        let is_host = {
+            let host = state.get_host();
+            host.is_some() && host.unwrap() == peer_id
+        };
+
+        // The state machine for the data channel established for this websocket.
+        while let Some(request) = receiver.next().await {
+            let request = match parse_request(request) {
+                Ok(request) => request,
+                Err(e) => {
+                    match e {
+                        ClientRequestError::Axum(_) => {
+                            // Most likely a ConnectionReset or similar.
+                            warn!("Unrecoverable error with {peer_id:?}: {e:?}");
+                        }
+                        ClientRequestError::Close => {
+                            info!("Connection closed by {peer_id:?}");
+                        }
+                        ClientRequestError::Json(_) | ClientRequestError::UnsupportedType(_) => {
+                            error!("Error with request: {:?}", e);
+                            continue; // Recoverable error
+                        }
+                    };
+                    if is_host {
+                        state.reset();
+                        // Lifecycle event: On Host Disonnected
+                        callbacks.on_host_disconnected.emit(peer_id);
+                    } else {
+                        state.remove_client(&peer_id);
+                        // Lifecycle event: On Client Disonnected
+                        callbacks.on_client_disconnected.emit(peer_id);
+                    }
+                    return;
+                }
+            };
+
+            match request {
```

Comment:
This block will probably be the same in 98% of cases, is there a way we could deduplicate it? Or provide helpers that make it shorter?